feat(developer): kmc-convert:  create a seperate util module for kmc-convert 😎

SabineSIL opened the pull request:
While developing kmc-convert a new set of functions have been added 

- Those should be moved to their own module.
- Tests should also be added for all those functions

see requests: 

- [JavaScript identifiers,](https://github.com/keymanapp/keyman/pull/12564#discussion_r2284361799)

- [use regex](https://github.com/keymanapp/keyman/pull/12564#discussion_r2284364419)
- [use regex](https://github.com/keymanapp/keyman/pull/12564#discussion_r2284365036)
- [use regex](https://github.com/keymanapp/keyman/pull/12564#discussion_r2284366120)
- [replace substring()](https://github.com/keymanapp/keyman/pull/12564#discussion_r2284367346)
- [Intl.Segmenter()](https://github.com/keymanapp/keyman/pull/12564#discussion_r2284368726)
- [replace indexOf()](https://github.com/keymanapp/keyman/pull/12564#discussion_r2284373674)

@keymanapp-test-bot skip
@Build-bot: skip release:developer
see PR #12564
see PR #14666

SabineSIL committed:
feat(developer): change var name hex_Entity->hexEntity

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): use ProcesData instead old convert_object

SabineSIL committed:
feat(developer): move convertToUnicodeCharacter etc to local module kmcConvertUtil.ts

SabineSIL committed:
feat(developer): new kmcConvertUtil module and tests moved from util.ts and kmn-file-writer.tests.ts

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): convertToUnicodeCodePoint, convertToUnicodeCharacter + tests to own kmcConvertUtil.ts/kmcConvertUtil.tests.ts

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
chore(developer): edits in function convertToUnicodeCodePoint

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): dummy change to start TC build

SabineSIL committed:
feat(developer): first changes in convertToUnicodeCharacter

SabineSIL committed:
feat(developer): changes in convertToUnicodeCharacter

SabineSIL committed:
feat(developer): move functions convertToUnicodeCodePoint and convertToUnicodeCharacter to convert-util

SabineSIL committed:
feat(developer): start rework function convertToUnicodeCharacter

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): finish rework of convertToUnicodeCharacter(); edit tests

SabineSIL committed:
feat(developer): comment out ToDo msg; comment out test

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): use const instead of let in utils.ts

SabineSIL committed:
feat(developer): edit tests for convertToUnicodeCharacter

SabineSIL committed:
feat(developer): edit tests for convertToUnicodeCharacter

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): enable build schema in build.sh

SabineSIL committed:
Merge remote-tracking branch 'origin/epic/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): use ERROR_UnsupportedCharactersDetected instead of INFO_UnsupportedCharactersDetected
  add tests for use of Unsupported output characters, add testfiles

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): remove writetoFile(); edit Error_UnsupportedCharactersDetected

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

# Conflicts:
#	developer/src/kmc-convert/src/keylayout-to-kmn/keylayout-to-kmn-converter.ts
#	developer/src/kmc-convert/src/keylayout-to-kmn/kmn-file-writer.ts

SabineSIL committed:
feat(developer): new convertControlCharacterToUnicodeCodePoint; add tests for convertControlCharacterToUnicodeCodePoint()

SabineSIL committed:
feat(developer): new convertControlCharacterToUnicodeCodePoint; add tests for convertControlCharacterToUnicodeCodePoint()

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): changed 1 test

SabineSIL committed:
feat(developer): remove convert-utils2.tests.ts

SabineSIL committed:
feat(developer): dummy change to start TC

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): change Info_UnsupportedCharactersDetected to ERROR_UnsupportedCharactersDetected
as in PR 14569 it is ERROR... whereas in PR 12564 it is still INFO...

SabineSIL committed:
feat(developer): resolve conflicts step 1

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): resolve conflict (symbol names; eroor/info_messages)

SabineSIL committed:
feat(developer): remove unneccessary else block; rename 1 symbol(inputFilename)

SabineSIL committed:
feat(developer): dummy change to start TC

SabineSIL committed:
feat(developer): dummy change to start TC

SabineSIL committed:
Merge branch 'master' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): dummy change to start TC

SabineSIL committed:
feat(developer): functional changes to remove convertCharacterToUnicodeCodePoint()

SabineSIL committed:
feat(developer): remove convertCharacterToUnicodeCodePoint()

SabineSIL committed:
feat(developer): dummy change to start TC ( add Build-bot: skip release:developer in PR descr.)

SabineSIL committed:
feat(developer): dummy change to start TC

SabineSIL committed:
feat(developer): disable ci_publish_npm

SabineSIL committed:
feat(developer): remove comments

SabineSIL committed:
feat(developer): remove function convertCharacterToUnicodeCodePoint()

SabineSIL committed:
feat(developer): remove tests

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
Merge branch 'epic/kmc-convert' into feat/developer/kmc-convert

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
Merge branch 'epic/kmc-convert' into feat/developer/kmc-convert

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): remove conflicts

SabineSIL committed:
feat(developer): change 1 test

SabineSIL committed:
feat(developer): edit name of test

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

# Conflicts:
#	developer/src/kmc-convert/test/keylayout-to-kmn-converter.tests.ts

SabineSIL committed:
feat(developer): dummy change to start TC

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

# Conflicts:
#	developer/src/kmc-convert/src/keylayout-to-kmn/keylayout-to-kmn-converter.ts
#	developer/src/kmc-convert/src/keylayout-to-kmn/kmn-file-writer.ts
#	developer/src/kmc-convert/test/keylayout-to-kmn-converter.tests.ts
#	developer/src/kmc-convert/test/kmn-file-writer.tests.ts

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): move convertToUnicodeCharacter() and tests to common
    convertToUnicodeCharacter()       ->  common\web\types\src\util\convert-utils.ts
    convertToUnicodeCharacter() tests ->  common\web\types\tests\util\convert-utils.tests.ts

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): remove some comments

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): remove duplicate tests change message text

SabineSIL committed:
feat(developer): adapt convertToUnicodeCharacter() with tests:
adapt use of unicode in .keylayout files and use html(hex) instead
ignore U+xxxx as output character in keylayout file (it is treated as string now)
add testfiles

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): resolve conflict :'key' does not exist in type ...

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

# Conflicts:
#	developer/src/kmc-convert/src/keylayout-to-kmn/keylayout-to-kmn-converter.ts

SabineSIL committed:
feat(developer): resolve conflict - convert parameter to Number

SabineSIL committed:
feat(developer): convert parameter to Number

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): resolve conflict - remove commented line

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): rename convertBound to unitTestEndpoints

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-Typesafety

# Conflicts:
#	developer/src/kmc-convert/src/keylayout-to-kmn/keylayout-to-kmn-converter.ts
#	developer/src/kmc-convert/src/keylayout-to-kmn/kmn-file-writer.ts
#	developer/src/kmc-convert/test/keylayout-to-kmn-converter.tests.ts
#	developer/src/kmc-convert/test/kmn-file-writer.tests.ts

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): add tests

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): remove test Test_characters.keylayout for now

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil

mcdurdin commented on common/web/types/src/util/convert-utils.ts:
```diff
@@ -0,0 +1,92 @@
+//***************************************************** */
+
```
```suggestion
```

mcdurdin commented on common/web/types/src/util/convert-utils.ts:
```diff
@@ -0,0 +1,92 @@
+//***************************************************** */
+
+/*
+ * Keyman is 2025 copyright (C) SIL International. MIT License.
```
Can you update to the new header style? https://github.com/keymanapp/keyman/wiki/Keyman-Code-Style-Guide#file-header

mcdurdin commented on common/web/types/src/util/convert-utils.ts:
```diff
@@ -0,0 +1,204 @@
+ *         a converted character if input is a valid Unicode value, numeric character reference in hex or decimal, or named character reference
+ *         or undefined if input is null or undefined, half a surrogate pair, or not recognized
+ */
+export function convertToUnicodeCharacter(inputString: string): string | undefined {
```
concluded that it can't, really

mcdurdin commented on developer/src/kmc-convert/src/keylayout-to-kmn/keylayout-to-kmn-converter.ts:
whitespace only, please remove from this PR

SabineSIL replied:
done

mcdurdin commented on developer/src/kmc-convert/test/data/Test_Character_Codepoint_C0.keylayout:
```diff
@@ -0,0 +1,91 @@
+            <key code="2" output="&#x42;"/>         <!-- C0-> html hex entity -->
+            <key code="3" output="&#67;"/>          <!-- C0-> html dec entity -->
+            <key code="4" output="&#x0044;"/>         <!-- C0-> unicode -->
+            <key code="5" output="&"/>              <!-- C0-> special: ampersand -->
```
This is not valid XML; `&` must be represented as `&amp;`.

SabineSIL replied:
done/removed

mcdurdin commented on developer/src/kmc-convert/test/data/Test_Character_Codepoint_C0.keylayout:
```diff
@@ -0,0 +1,91 @@
+            <key code="2" output="&#x1250;"/>                  
+            <key code="3" output="&#4680;"/>        
+            <key code="4" output="&#x1244;"/>
+            <key code="5" output="&&"/>              <!-- C0-> special: double ampersand -->
```
here too

SabineSIL replied:
done/removed

mcdurdin commented on developer/src/kmc-convert/test/data/Test_Character_Codepoint_C0.keylayout:
```diff
@@ -0,0 +1,91 @@
+            <key code="2" output="&#x1F60E;"/>      
+            <key code="3" output="&#128508;"/>      
+            <key code="4" output="&#x1F61E;"/>
+            <key code="5" output="&;"/>              <!-- C0-> special: ampersand ; -->
```
here too

SabineSIL replied:
done/removed

mcdurdin commented on developer/src/kmc-convert/test/data/Test_Character_Codepoint_C0.keylayout:
```diff
@@ -0,0 +1,91 @@
+            <key code="2" output="&#x1E99;"/>       
+            <key code="3" output="&#7844;"/>        
+            <key code="4" output="&#x1E99;"/>
+            <key code="5" output="&&;"/>             <!-- C0-> special: double ampersand ; -->
```
here too

SabineSIL replied:
done/removed

mcdurdin commented on developer/src/kmc-convert/test/data/Test_Character_Codepoint_C3.keylayout:
```diff
@@ -0,0 +1,61 @@
+            <when state="1" output="â"/>
+        </action>
+        <action id="A7">
+            <when state="1" output="&#0007;"/>
```
```suggestion
            <!-- this is not technically valid XML but .keylayout files include C0 characters -->
            <when state="1" output="&#0007;"/>
```

SabineSIL replied:
done

mcdurdin commented on developer/src/kmc-convert/test/data/Test_characters.keylayout:
```diff
@@ -24,20 +24,25 @@
-            <key code="3" output="U+1F600"/>
-            <key code="4" output="U+0002"/>
+            <key code="3" output="&#x1F600;"/>
+            <key code="4" output="&#x0002;"/>
```
```suggestion
            <!-- this is not technically valid XML but .keylayout files include C0 characters -->
            <key code="4" output="&#x0002;"/>
```

SabineSIL replied:
done

mcdurdin commented on VERSION.md:
Please remove this file from the PR

SabineSIL replied:
done

mcdurdin commented on developer/src/kmc-convert/build.sh:
```diff
@@ -46,5 +46,6 @@ do_test() {
   builder_echo warning "Please increase threshold in build.sh as test coverage improves."
 }
 
+#builder_run_action test        do_kmc_convert_test  # build schema,...
```
```suggestion
```

SabineSIL replied:
done

mcdurdin commented on developer/src/kmc-convert/build.sh:
```diff
@@ -46,5 +46,6 @@ do_test() {
+#builder_run_action test        do_kmc_convert_test  # build schema,...
 builder_run_action test        do_test
-builder_run_action publish     ci_publish_npm
+#builder_run_action publish     ci_publish_npm
```
We will need this for release

SabineSIL replied:
done/removed

SabineSIL replied:
<img width="752" height="82" alt="image" src="https://github.com/user-attachments/assets/a60b821e-bc97-4296-ba42-4afeddefc976" />

SabineSIL committed:
feat(developer): kmc-convert review comments of PR 14569
 - info html control characters
 - remove whitespace
 - file header
 - add builder_run_action publish ci_publish_npm in build.sh

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

# Conflicts:
#	developer/src/kmc-convert/src/keylayout-to-kmn/keylayout-to-kmn-converter.ts

SabineSIL committed:
feat(developer): comment out builder_run_action publish

mcdurdin commented on common/web/types/src/util/convert-utils.ts:
```diff
@@ -0,0 +1,90 @@
+    }
+    // valid '&gt', '&lt',..
+    else if (m_nam) {
+        switch (m_nam[1].toLowerCase()) {
```
`toLowerCase()` is not required -- these are case-sensitive.

SabineSIL replied:
done

mcdurdin commented on common/web/types/src/util/convert-utils.ts:
```diff
@@ -0,0 +1,90 @@
+    const m_chr = /^(?!&).+$/i.exec(inputString);
+
+    // '&', '&#','&#x' with or without ; will later be refused for conversion
+    const m_chr_inv = /^((&;?)+|(&#;?)+|(&#x;?)+;?)$|^$/i.exec(inputString);
```
It would be cleaner to define the regexes in a class, so that they are instantiated and compiled just once. Something like this:

```ts
class UnicodeCharacterConversion {
  // &#x followed by 1.-6. hex digits will later be used for conversion
  private static re_hex = /^&#x([0-9a-f]{1,6});$/i;

  // &# followed by 1.-6. decimal digits will later be used for conversion
  private static re_dec = /^&#([0-9]{1,7});$/;

  ...

  public convert(inputString: string): string | undefined {
    const m_hex = UnicodeCharacterConversion.re_hex.exec(inputString);
    if(m_hex) {
      const codePoint_h = parseInt(m_hex[1], 16);
      // Reject surrogates and invalid codepoints
      if ((codePoint_h >= 0xD800 && codePoint_h <= 0xDFFF) || codePoint_h > 0x10FFFF) {
          return undefined;
      }
      return String.fromCodePoint(codePoint_h);      
    }

    const m_dec = ...
  }
};
```

SabineSIL replied:
done

mcdurdin commented on common/web/types/src/util/convert-utils.ts:
```diff
@@ -0,0 +1,90 @@
+        // Reject surrogates and invalid codepoints
+        if ((codePoint_h >= 0xD800 && codePoint_h <= 0xDFFF) || codePoint_h > 0x10FFFF) {
+            return undefined;
+        }
```
Test for valid characters should use functions from common\web\types\src\util\util.ts (util in `@keymanapp/common-types`)

SabineSIL replied:
done

mcdurdin commented on common/web/types/src/util/convert-utils.ts:
```diff
@@ -0,0 +1,90 @@
+ *         a converted character if input is a valid Unicode value, numeric character reference in hex or decimal, or named character reference
+ *         or undefined if input is null or undefined, half a surrogate pair, or not recognized
+ */
+export function convertToUnicodeCharacter(inputString: string): string | undefined {
```
I don't think this function will work correctly for strings like: `a&gt;b`?

If an arbitrary-length string can be passed in, then we need to scan the string for every instance of `&` and validate.

SabineSIL replied:
In this case the function will return  `a&gt;b` :

<img width="1086" height="197" alt="image" src="https://github.com/user-attachments/assets/ed4e4c0c-4902-4012-b6ac-2a0e150fc68e" />


<img width="262" height="70" alt="image" src="https://github.com/user-attachments/assets/c296284f-e26c-4aa3-9d49-b99ad5f69024" />

mcdurdin replied:
We need to iterate on `UnicodeCharacterConversion.convert` a bit more -- when I look at the usage, it's called _only_ from `writeCharacterOrUnicode` in `KmnFileWriter`, which itself _also_ declares regexes for character entities (and `U+xxxx`?). So there's some double-processing going on and it seems like it might be overly complicated? Furthermore, it looks like these functions are _only_ being used for reporting errors as comments in the output file?

As that is the case, I would like to pull this _out_ of common/web/types and put it alongside kmn-file-writer.ts, and thus keep its usage constrained to this module. That reduces the need to iterate on its design further.

Otherwise... I'm not comfortable with the function as it stands -- sending it a string with multiple characters is a failure mode for the function which will lead to bugs in the future. It should either work with a single codepoint, or with a whole string. If a single codepoint, then it should chop off the section it processes from the beginning of the string, and return the remainder of the string for further conversion.

SabineSIL replied:
A lot of proposals have been made to move this function-moved it back to kmn-file-writer.ts. I hope it is the correct location now.

> So there's some double-processing going on and it seems like it might be overly complicated? 

- only 4 regexes are used from class UnicodeCharacterConversion now

> it looks like these functions are only being used for reporting errors as comments in the output file?

- No it is not only used for writing messages. It can but it is used mainly to process characters and html entities
 ( e.g `<key code="1" output="&gt;"/>  <!-- S -->` which writes `+ [NCAPS K_S]  >  '>' ` )

<img width="1256" height="865" alt="image" src="https://github.com/user-attachments/assets/4ef5861c-2dd8-4686-958b-ff99601adab5" />



A few changes have been made to the function: 

- in kmn-file-writer: renamed convert() to convert_htmlToCharacter() for clarity since in keylayourt-to-kmn-converter there is already a convert() function
- new function unescape_string() to unescape (double) encoded named html entities  ( &gt ;  &amp ;amp;)
- a new recursive function processXmlValue to processes each character/html entity in an input string 
writeCharacterOrUnicode

mcdurdin commented on developer/src/kmc-convert/src/keylayout-to-kmn/keylayout-to-kmn-converter.ts:
```diff
@@ -228,7 +228,17 @@ export class KeylayoutToKmnConverter {
           // ...............................................................................................................................
 
-         if (jsonObj.keyboard.keyMapSet[0].keyMap[i].key[j]['output'] !== undefined) {
+          if (jsonObj.keyboard.keyMapSet[0].keyMap[i].key[j]['output'] === "") {
+            /* this.callbacks.reportMessage(ConverterMessages.Warn_EmptyOutput({
+             keymapIndex: jsonObj.keyboard.keyMapSet[0].keyMap[i]['index'],
+             key: jsonObj.keyboard.keyMapSet[0].keyMap[i].key[j]['code'],
+             KeyName: this.mapUkeleleKeycodeToVK(Number(jsonObj.keyboard.keyMapSet[0].keyMap[i].key[j]['code']))
+           }));
+           return null;*/
```
Commented code?

SabineSIL replied:
this (and a few other comments) will stay there until the final merge is done

mcdurdin replied:
Can you add `// TODO-KMC-CONVERT: remove` comments for each of these? Makes it easier to audit prior to merge

SabineSIL replied:
added` // TODO-KMC-CONVERT: remove` even though there was an almost identical comment already

this loop is necessary to write a Warning if the output character is '' which might be created by fast-xml-parser for a non valid control character in htlml:

**keylayout**: 
`<key code="3" output="&#0001;"/>  <!-- F -->`
**kmn**: 
`c WARNING: empty output or unsupported numerical html entity here: + [CAPS K_F]  >  ''`

SabineSIL replied:
added even though there have been similar comments

SabineSIL committed:
feat(developer): use util.isValidUnicode() to check codePoints; remove ci_publish_npm

SabineSIL committed:
feat(developer): define regexes in class

SabineSIL committed:
feat(developer): remove VERSION.md from git

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): remove VERSION.md from git again

SabineSIL committed:
feat(developer): remove commented code

mcdurdin committed:
chore: restore VERSION.md

mcdurdin commented on common/web/types/src/util/convert-utils.ts:
Should this function be deleted now? It is referenced in a couple of comments and then in  a unit test which needs to be updated to point to the new class function

mcdurdin commented on common/web/types/src/util/convert-utils.ts:
```diff
@@ -0,0 +1,173 @@
+    private static re_chr_inv = /^((&;?)+|(&#;?)+|(&#x;?)+;?)$|^$/i;
+
+
+    public convert(inputString: string): string | undefined {
```
```suggestion
    public static convert(inputString: string): string | undefined {
```

Then we do not need to instantiate the class

mcdurdin commented on developer/src/kmc-convert/src/keylayout-to-kmn/kmn-file-writer.ts:
```diff
@@ -1772,7 +1775,8 @@ export class KmnFileWriter {
     else {
-      out.character = this.convertToUnicodeCharacter(ctr);;
+      const uni_conv = new convertUtil.UnicodeCharacterConversion();
+      out.character = uni_conv.convert(ctr) ?? "";
```
```suggestion
      out.character = convertUtil.UnicodeCharacterConversion.convert(ctr) ?? "";
```

SabineSIL replied:
done and changed function name to  `UnicodeCharacterConversion.processXmlValue()`

mcdurdin reviewed (changes requested):
Can you refactor as requested and then we'll go through one last review?

mcdurdin commented on developer/src/kmc-convert/src/keylayout-to-kmn/kmn-file-writer.ts:
Suggest that we add a note here that this is intended to be replaced with the AST version once ng-compiler is complete, so we remember that later!

SabineSIL replied:
had been stated in the document also

SabineSIL replied:
added

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): enable 'use of unsupported numerical html entity' message again

SabineSIL committed:
feat(developer): rewrite kmn-file-writer-convert():
now use convert_htmlToCharacter, unescape_string, processXmlValue

SabineSIL committed:
feat(developer): remove convet_utils and convert_utils.tests

SabineSIL committed:
feat(developer): remove convet_utils and convert_utils.tests

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): add message for use of Unicode codepoints; edit regex, unescape_string; handle empty input

SabineSIL committed:
feat(developer): add message for use of Unicode codepoints; edit regex, unescape_string; handle empty input

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert-createSeperateUtil' of https://github.com/keymanapp/keyman into feat/developer/kmc-convert-createSeperateUtil

SabineSIL committed:
feat(developer): add outputfilename for empty name

SabineSIL committed:
feat(developer): reviewRules edit typo in use of array( amb_4_1[0].idDeadkey instead amb_2_1[0].idDeadkey)

SabineSIL committed:
feat(developer): handle empty outputfilename removed; processXmlValue: handled inputString.replaced_character === undefined

SabineSIL committed:
Merge branch 'feat/developer/kmc-convert' into feat/developer/kmc-convert-createSeperateUtil
